bug 1328989: Add /en-US/docs/macros for searching

jwhitlock opened the pull request:
Display a table of macro names and page counts in English and all documents, and include a form for manual macro searches for macros not known to KumaScript.

jwhitlock commented:
I could use other's opinions on some follow-on items:

* Does this belong at /docs/macros, or /dashboards/macros?  I have it at /docs/macros because that's where docs/templates are, but "special" doc pages seem out of place to me.
* Should I drop /docs/templates in this PR as well? If so, should I use a standard wiki redirect, or a URL rule to redirect?
* Should I cache the macro and usage lists for the view, or wait until we see the performance in prod?

stephaniehobson commented:
> Does this belong at /docs/macros, or /dashboards/macros? I have it at /docs/macros because that's where docs/templates are, but "special" doc pages seem out of place to me.

dashboards/macros

> Should I drop /docs/templates in this PR as well? If so, should I use a standard wiki redirect, or a URL rule to redirect?

yes, ummmm, standard wiki redirect?

> Should I cache the macro and usage lists for the view, or wait until we see the performance in prod?

dunno

escattone commented:
> Does this belong at /docs/macros, or /dashboards/macros? I have it at /docs/macros because that's where docs/templates are, but "special" doc pages seem out of place to me.

dashboards/macros

> Should I drop /docs/templates in this PR as well? If so, should I use a standard wiki redirect, or a URL rule to redirect?

Yes. I'm not clear about the pros/cons of each choice, but might we do a 301 via the Django view?

> Should I cache the macro and usage lists for the view, or wait until we see the performance in prod?

I'm loading the anonymized DB to check with more data, but my current feeling is that we can wait to see the performance in production.

stephaniehobson commented:
I was thinking standard wiki redirect because it would be less code in kuma. But now I'm realizing that we won't be able to redirect the entire sub tree, just the landing page. So mostly I stand by my 'ummm'

safwanrahman commented on kuma/wiki/jinja2/wiki/list/macros.html:
```diff
@@ -0,0 +1,71 @@
+    <label for="name">{{ _("Search for macro usage:") }}</label>
+    <input id="name" name="kumascript_macros" type="text" placeholder="{{_("Macro Name") }}" />
+    <br/>
+    <button type="submit" name="locale" value="*">{{ ("Search All Pages") }}</button>
```
This text should be unlocalized for any reason?

jwhitlock replied:
good catch, should be localized

jwhitlock commented:
I'm asking for @stephaniehobson's review as well because she's the expert on displaying tabular data.  Follow-on commits coming to:

* [ ] Localize buttons
* [ ] Move the page to /en-US/dashboards/macros
* [ ] Drop /en-US/docs/templates

safwanrahman commented on kuma/wiki/tests/test_kumascript.py:
```diff
@@ -60,3 +81,166 @@ def test_get_does_not_call_format_slug_for_docs(self, mock_format_slug):
+        'A11yRoleQuicklinks': 'A11yRoleQuicklinks.ejs',
+        'APIFeatureList': 'APIFeatureList.ejs',
+    }
+    assert macros == expected
```
I would prefer using ``eq_``
It returns clear output if fail

jwhitlock replied:
Please post the output of ``eq_`` vs ``assert`` in a failing test.

safwanrahman replied:
Reading source, seems like its deprecated
https://github.com/mozilla/kuma/blob/master/kuma/core/tests/__init__.py#L17

escattone replied:
@safwanrahman We shouldn't use the ``eq_``, ``ok_``, etc. helpers any longer, as pytest takes care of providing detailed information on failure using the standard ``assert``.

safwanrahman commented on kuma/wiki/tests/test_templates.py:
```diff
@@ -1331,3 +1331,46 @@ def _translation_data():
+
+    response = client.get(reverse('wiki.list_macros'), follow=True)
+    assert response.status_code == 200
+    assert "Found 1 active macro." in response.content
```
can use ``ok_``

escattone commented:
Thanks @jwhitlock, this is an awesome PR! I'm really happy to see the docs/templates updated to fit the new reality for macros.

So far, I've found what appear to be two issues:
- If I click on any number link in the "All Page Count" column (or do a manual "Search All Pages"), the search results header always says "...documents found in English (US)" but the results often contain links to non-English documents as well. It seems it should say something like "...documents found in all locales"?
- If I do a manual search (which is inactivated once one of the search buttons is clicked), view the results, then hit the browser's back button, the manual search remains inactivated.

stephaniehobson commented on kuma/wiki/jinja2/wiki/list/macros.html:
```diff
@@ -0,0 +1,71 @@
+    <input id="name" name="kumascript_macros" type="text" placeholder="{{_("Macro Name") }}" />
+    <br/>
+    <button type="submit" name="locale" value="*">{{ ("Search All Pages") }}</button>
+    <button type="submit" name="locale" value="en-US">{{ ("Search English Pages") }}</button>
```
Needs localization underscore here too.

stephaniehobson commented on kuma/wiki/jinja2/wiki/list/macros.html:
```diff
@@ -0,0 +1,71 @@
+      </tbody>
+    </table>
+  {% endif %}
+  <form id="macro-search" method="GET" action="{{url("search")}}">
```
Please add `class="nodisable"`. This will stop the form from being disabled when using the back button to return to the page after a search.

escattone commented on tests/functional/test_lists.py:
```diff
@@ -0,0 +1,15 @@
+
+
+@pytest.mark.nondestructive
+def test_list_macros(base_url, selenium):
```
I'm wondering if we want to more fully check the functionality of this page, like adding assertions and/or click-tests for the links (macro-name, macro-counts) and the manual search buttons?

escattone commented:
Thanks for the changes @jwhitlock! There's still this issue I'd like to highlight again and get everyone's feedback.

- For the case when we're searching all locales, for example if I click on a number link in the "All Page Count" column (or do a manual "Search All Pages"), the search results page always says "...documents found in English (US)" but it seems to me it should say something like "...documents found in all locales" or simply "...documents found"? (https://github.com/mozilla/kuma/blob/master/kuma/search/jinja2/search/results.html )

jwhitlock commented:
@escattone yes, I think the template needs to reflect when all pages are searched.  I'd like to do that in a different PR.

escattone commented on kuma/wiki/kumascript.py:
```diff
@@ -298,7 +299,11 @@ def macro_sources():
-        return {md['name']: md['filename'] for md in macros_raw}
+        # Ensure Normal Form C used on GitHub
+        return {
+            unicodedata.normalize('NFC', md['name']):
```
Wow, I didn't even realize there were normalization forms for unicode. Good catch and nice work @jwhitlock !

jwhitlock replied:
I didn't either, until I hit issues exporting [historical macros to files](https://github.com/mdn/archived_kumascript/blob/master/export_templates.py#L162-L176).

jwhitlock commented:
I'm done with this round of changes:

* Moved from ``docs/macros`` to ``dashboards/macros``
* Localized manual search buttons, added ``class='no disable'``, and added element IDs for testing
* Made ``docs/templates`` a permanent redirect to ``dashboards/macros``
* Fixed a unicode normalization issue when there are two ways of encoding a character.  For example, the `ó` in "traducción.ejs" can be encoded as two characters, "o", and then a "combined with a diacritical", with the invalid link https://github.com/mozilla/kumascript/blob/master/macros/traduccio%CC%81n.ejs. It can also be encoded as "O WITH ACUTE, LATIN SMALL LETTER", with the valid link https://github.com/mozilla/kumascript/blob/master/macros/traducci%C3%B3n.ejs.
* Remove a duplicate call to ``macro_sources``
* Added unit test for an unexpected ElasticSearch error on the second call
* Added functional tests that follow GitHub and search links, and exercise manual search
* Add a permanent redirect from ``docs/templates`` to ``dashboards/macros``.

I did not add a generic redirect for template source, like [docs/Template:a11yEventRef](https://developer.mozilla.org/en-US/docs/Template:a11yEventRef). Due to MDN's flexible capitalization rules, I don't think there is a reliable way to link to the new source.  I think these should be 404s when the template source is removed from MDN, and any MDN links to the files should be updated to GitHub (or maybe the MDN docs moved to the GitHub repo).

I also didn't fix the misleading text "documents found in English (US)" when searching all locales, since I feel this is an existing issue that should be addressed with a new PR.

escattone commented:
Wow, thanks again @jwhitlock, this is awesome work with really thorough well-written tests!

jwhitlock commented:
Found [bug 1063560](https://bugzilla.mozilla.org/show_bug.cgi?id=1063560) - "documents found in English." is a 3 year-old bug 😠
